rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -148,29 +148,32 @@ class Publisher : public PublisherBase
   {
     // Topic is unused for now.
     (void)topic;
+    (void)qos;
     (void)options;
 
     // If needed, setup intra process communication.
     if (rclcpp::detail::resolve_use_intra_process(options_, *node_base)) {
       auto context = node_base->get_context();
       // Get the intra process manager instance for this context.
       auto ipm = context->get_sub_context<rclcpp::experimental::IntraProcessManager>();
-      // Register the publisher with the intra process manager.
-      if (qos.history() != rclcpp::HistoryPolicy::KeepLast) {
+      // Check if the QoS is compatible with intra-process.
+      auto qos_profile = get_actual_qos();
+      if (qos_profile.history() != rclcpp::HistoryPolicy::KeepLast) {
         throw std::invalid_argument(
                 "intraprocess communication allowed only with keep last history qos policy");
       }
-      if (qos.depth() == 0) {
```

Comment:
this was the direct cause of https://github.com/ros2/rclcpp/issues/2705, QoS via the argument is NOT the actual QoS from rmw implementation with SystemDefaultQoS(this means rmw implementation default). So here it just uses `RMW_QOS_POLICY_DEPTH_SYSTEM_DEFAULT`, which is 0, and eventually causes this exception.

by the way, subscription does not have this problem since it already calls `get_actual_qos()` for IPC subscriber. this is why we do not have the problem on creation of IPC subscription.